### PR TITLE
Fix race condition in HIP backend

### PR DESCRIPTION
**Pull request by Rombur**
This PR add new `__syncthreads()` after shuffle and move another synchronization just after a shuffle. This allows us to enable one more tests. This should not be necessary but there is a race condition in shuffle with HIP. This bug has already been reported to AMD.

### CODE REVIEW

File: core/src/HIP/Kokkos_HIP_Team.hpp [hunk 1]
```diff
@@ -297,6 +300,9 @@ class HIPTeamMember {
     for (int i = blockDim.x; (i >>= 1);) {
       ::Kokkos::Experimental::Impl::in_place_shfl_down(tmp2, tmp, i,
                                                        blockDim.x);
+      // FIXME_HIP the synchronization point is necessary because of a race
+      // condition in shuffle
+      __syncthreads();
```
Comment (crtrott) [inline core/src/HIP/Kokkos_HIP_Team.hpp:305]: you can't do this. Not in a vector_reduce.

Comment (crtrott) [reply core/src/HIP/Kokkos_HIP_Team.hpp:305]: i.e. this will not be called by all threads in the team .

Comment (crtrott) [reply core/src/HIP/Kokkos_HIP_Team.hpp:305]: did you try using a memory fence instead?

File: core/src/HIP/Kokkos_HIP_Team.hpp [hunk 2]
```diff
@@ -308,6 +314,9 @@ class HIPTeamMember {
     // and thus different threads could have different results.
 
     ::Kokkos::Experimental::Impl::in_place_shfl(tmp2, tmp, 0, blockDim.x);
+    // FIXME_HIP the synchronization point is necessary because of a race
+    // condition in shuffle
+    __syncthreads();
```
Comment (crtrott) [inline core/src/HIP/Kokkos_HIP_Team.hpp:319]: can't do this in vector_reduce.

File: core/src/HIP/Kokkos_HIP_Team.hpp [hunk 3]
```diff
@@ -969,6 +981,9 @@ KOKKOS_INLINE_FUNCTION void parallel_scan(
     for (int j = 1; j < static_cast<int>(blockDim.x); j <<= 1) {
       value_type tmp = 0;
       ::Kokkos::Experimental::Impl::in_place_shfl_up(tmp, sval, j, blockDim.x);
+      // FIXME_HIP the synchronization point is necessary because of a race
+      // condition in shuffle
+      __syncthreads();
```
Comment (crtrott) [inline core/src/HIP/Kokkos_HIP_Team.hpp:986]: can't do this in ThreadVectorRange

File: core/src/HIP/Kokkos_HIP_Team.hpp [hunk 4]
```diff
@@ -983,6 +998,9 @@ KOKKOS_INLINE_FUNCTION void parallel_scan(
     // Accumulate the last value in the inclusive scan:
     ::Kokkos::Experimental::Impl::in_place_shfl(sval, sval, blockDim.x - 1,
                                                 blockDim.x);
+    // FIXME_HIP the synchronization point is necessary because of a race
+    // condition in shuffle
+    __syncthreads();
```
Comment (crtrott) [inline core/src/HIP/Kokkos_HIP_Team.hpp:1003]: can't do this in ThreadVectorRange

File: core/src/HIP/Kokkos_HIP_Team.hpp [hunk 5]
```diff
@@ -170,6 +170,9 @@ class HIPTeamMember {
       ::Kokkos::Experimental::Impl::in_place_shfl(
           val, tmp, blockDim.x * thread_id, blockDim.x * blockDim.y);
     }
+    // FIXME_HIP the synchronization point is necessary because of a race
+    // condition in shuffle
+    __syncthreads();
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Team.hpp:175]: Why did you not just put the synchronization point in the shuffle function?

File: core/src/HIP/Kokkos_HIP_Team.hpp [hunk 6]
```diff
@@ -1025,6 +1043,9 @@ KOKKOS_INLINE_FUNCTION void single(
 #ifdef __HIP_DEVICE_COMPILE__
   if (threadIdx.x == 0) lambda(val);
   ::Kokkos::Experimental::Impl::in_place_shfl(val, val, 0, blockDim.x);
+  // FIXME_HIP the synchronization point is necessary because of a race
+  // condition in shuffle
+  __syncthreads();
```
Comment (crtrott) [inline core/src/HIP/Kokkos_HIP_Team.hpp:1048]: can't do this in PerThread single struct

**Review by crtrott — CHANGES_REQUESTED** [review]
you can't call __syncthreads inside of vector level operations. Maybe a threadfence will work instead? Or we could try a local __shared variable reduction instead of shfl?

**Review by dalg24 — COMMENTED** [review]
How about https://github.com/kokkos/kokkos/blob/4231f80c67c84697dc1e6d889705ea69aa45fa56/core/src/HIP/Kokkos_HIP_ReduceScan.hpp#L85

File: core/src/HIP/Kokkos_HIP_Team.hpp (hunk 5 above)
Comment (crtrott) [reply core/src/HIP/Kokkos_HIP_Team.hpp:175]: again this doesn't all doesn't work since these functions are not block synchronous called ...

**Comment by Rombur** [comment]
I have reworked the PR. I have added `__threadfence()` after the call to `__shfl` and remove the synchronizations that are not needed.

**Review by crtrott — APPROVED** [review]

**Pull request merged**